R/data_color.R

```diff
@@ -391,131 +393,118 @@ adjust_luminance <- function(colors,
 
-      } else if (color %>% is_hex_col_w_alpha()) {
+  # Normalize color to a #RRGGBB (stripping the alpha channel)
+  bgnd_color <- process_color(colors = bgnd_color, alpha = 1)
 
-        color %>%
-          substring(1, 7) %>%
-          toupper()
-
-      } else if (tolower(color) %in% grDevices::colors()) {
-
-        color_matrix <- grDevices::col2rgb(tolower(color))
-
-        grDevices::rgb(
-          red = color_matrix[1, ]/255,
-          green = color_matrix[2, ]/255,
-          blue = color_matrix[3, ]/255
-        )
+  # Determine the ideal color for the chosen background color
+  yiq_contrasted_threshold <- 128
+  colors <- grDevices::col2rgb(bgnd_color)
+  score <- colSums(colors * c(299, 587, 144)) / 1000
+  ifelse(score >= yiq_contrasted_threshold, dark, light)
+}
 
-      } else {
-        NA_character_
-      }
-    })
+#' Convert a single color in the rgba() string format to hexadecimal (#RRGGBBAA)
+#'
+#' This function will accept a single color in the rgba() string format (e.g.,
+#' "`rgba(255,170,0,0.5)`") and output a color in the hexadecimal format with an
+#' alpha component (#RRGGBBAA).
+#'
+#' @noRd
+rgba_to_hex <- function(color) {
+
+  channels <-
+    color %>%
+    tidy_gsub(pattern = "(rgba\\(|\\))", "") %>%
+    strsplit(",") %>%
+    unlist() %>%
+    as.numeric()
+
+  grDevices::rgb(
+    red = channels[1] / 255,
+    green = channels[2] / 255,
+    blue = channels[3] / 255,
+    alpha = channels[4]
+  )
 }
 
-#' With any input color, transform to hexadecimal with alpha
+#' With a vector of input colors return normalized color strings
+#'
+#' Input colors can be color names (e.g., `"green"`, `"steelblue"`, etc.),
+#' colors in hexadecimal format with or without an alpha component (either
+#' #RRGGBB or #RRGGBBAA), or color strings in the rgba() string format
+#' (e.g., "`rgba(255,170,0,0.5)`"). Output is the same length vector as the
+#' input but it will contain a mixture of either #RRGGBB colors (if the input
+#' alpha value for a color is 1) or rgba() string format colors (if the input
+#' alpha value for a color is not 1).
 #'
 #' @noRd
-normalize_color <- function(colors,
-                            alpha) {
+process_color <- function(colors, alpha = NULL) {
 
-  if (length(colors) == 0) {
-    stop("One or more colors must be provided", call. = FALSE)
+  col_matrix_to_rgba <- function(color_matrix) {
+
+    color_matrix %>%
+      as.data.frame() %>%
+      dplyr::mutate(
+        rgba = paste0("rgba(", r, ",", g, ",", b, ",", round(alpha, 2), ")")
+      ) %>%
+      dplyr::pull(rgba)
   }
 
-  if (!(length(colors) == length(alpha) || length(alpha) == 1)) {
-    stop("The length of `alpha` must be the same length as `colors` or `1`",
-         call. = FALSE)
+  color_matrix <-
+    vapply(
+      colors,
+      FUN.VALUE = character(1),
+      USE.NAMES = FALSE,
+      FUN = function(color) {
+        if (is_rgba_col(color)) color <- rgba_to_hex(color = color)
```

Comment:
Just curious: is there any way we'd end up with `rgba()` colors here, besides the user manually typing in `rgba()` colors? I'm wondering if we even want to allow that syntax, or require people to provide gt with `#RRGGBB` or `#RRGGBBAA` colors (and `#RGB` if you're feeling generous). If there is a good reason to support `rgba()` (like if you might have to run this function on its own output??) then do we also need to support `rgb()`?

Reply:
The user should never provide rgba()-style colors. The reason this was included here is because a downstream function `ideal_fgnd_color()` will take the converted color (and it may be in rgba()) but requires a #RRGGBB color. I've since removed this `vapply()` and will instead use a revised `rgba_to_hex()` directly inside `ideal_fgnd_color()`.

I think this was a case of too much coupling.